File: 09-dictionaries/02_Hash_Methods/spec.rb
describe "merge_us" do
  it "merges two hashes that are unique" do
    h1 = { name: "Computer", cost: "$1,000" }
    h2 = { first_name: "Bob", age: 34 }
    new_hash = { name: "Computer", cost: "$1,000", first_name: "Bob", age: 34 }

    merge_us(h1, h2).should eq(new_hash)
  end
  it "merges two hashes that are have some things in common" do
    h1 = { name: "Computer", cost: "$1,000" }
    h2 = { name: "Mouse", uuid: "1234" }
    new_hash = { name: "Mouse", cost: "$1,000", uuid: "1234" }
    
    merge_us(h1, h2).should eq(new_hash)
  end
end

describe "my_keys" do
  it "returns keys for a small hash" do
    h = { name: "Computer", cost: "$1,000" }
    keys = [:name, :cost]

    my_keys(h).should eq(keys)
  end
  it "returns keys for a larger hash" do
    h = { name: "Mouse", cost: "$5", uuid: "1234" }
    keys = [:name, :cost, :uuid]

    my_keys(h).should eq(keys)
  end
end

describe "do_i_have?" do
  it "returns false if it doesn't have any of the keys" do
    h = { name: "Computer", cost: "$1,000" }
    keys = [:age, :bio]

    do_i_have?(h, keys).should eq(false)
  end
  it "returns false if one or more of the keys isn't in the hash" do
    h = { name: "Computer", cost: "$1,000" }
    keys = [:name, :bio, :cost]

    do_i_have?(h, keys).should eq(false)
  end
  it "returns false if the hash has a different number of keys than the array" do
    h = { name: "Computer", cost: "$1,000" }
    keys = [:name]

    do_i_have?(h, keys).should eq(false)
  end  
  it "returns true if all keys are in the hash" do
    h = { name: "Computer", cost: "$1,000", uuid: "1234" }
    keys = [:name, :cost, :uuid]

    do_i_have?(h, keys).should eq(true)
  end
  it "returns true if all keys are in the hash, regardless of order" do
    h = { name: "Computer", cost: "$1,000", uuid: "1234" }
    keys = [:name, :uuid, :cost]

    do_i_have?(h, keys).should eq(true)
  end
end

File: 10-loops/02-object-operations/spec.js
var expect = require('chai').expect;
var solution = require('./challenge');
var _ = require('underscore');

describe("copy", function() {
  it("should create a new object with a copy of all the data", function() {
    orig = {name: "Rob"};
    result = solution.copy(orig);
    expect(result).to.not.equal(orig);
    expect(result).to.deep.eql(orig);
  });
});

describe("extend", function() {
  it("should correctly extends two hashes that are unique", function() {
    dest = { name: "Computer", cost: "$1,000" };
    src = { first_name: "Bob", age: 34 };
    result = solution.extend(dest, src);

    expect(result).to.equal(dest, "Returned hash must be the same as the original destination hash.");
    expect(result).to.have.keys(['name', 'cost', 'first_name', 'age']);
  });

  it("should correctly extend two hashes that are have keys in common", function() {
    dest = { name: "Computer", cost: "$1,000" };
    src = { name: "Mouse", uuid: "1234" };
    result = solution.extend(dest, src);

    expect(result).to.have.keys(['name', 'cost', 'uuid']);
    expect(result).to.have.property('name', 'Mouse');
    expect(result).to.have.property('cost', '$1,000');
  });
});


describe("hasElems", function() {
  var h = {};
  beforeEach(function() {
    h = { name: "Computer", cost: "$1,000" };
  });

  it("returns true for an empty array", function() {
    expect(solution.hasElems(h, [])).to.be.true;
  });


  it("returns true if it has all the keys", function() {
    keys = ['name', 'name', 'name']
    expect(solution.hasElems(h, keys)).to.be.true;
  });

  it("returns false if one or more of the keys isn't in the hash", function() {
    var keys = ['name', 'bio', 'cost'];
    expect(solution.hasElems(h, keys)).to.be.false;
  });
});
